File: bindings/haskell/src/ProvenServers/Nesy.hs
-- SPDX-License-Identifier: MPL-2.0
-- Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
--
-- | NeSy types for the proven-servers ABI.
--
-- All tag values match the Idris2 ABI discriminants exactly.

module ProvenServers.Nesy
  (
    ReasoningMode(..)
  , reasoningModeToTag
  , reasoningModeFromTag
  , ProofStatus(..)
  , proofStatusToTag
  , proofStatusFromTag
  , ConstraintKind(..)
  , constraintKindToTag
  , constraintKindFromTag
  , NeuralBackend(..)
  , neuralBackendToTag
  , neuralBackendFromTag
  , Confidence(..)
  , confidenceToTag
  , confidenceFromTag
  , DriftKind(..)
  , driftKindToTag
  , driftKindFromTag
  , NeSyState(..)
  , neSyStateToTag
  , neSyStateFromTag
  ) where

import Data.Word (Word8)

-- ---------------------------------------------------------------------------
-- ReasoningMode
-- ---------------------------------------------------------------------------

-- | Neurosymbolic reasoning modes.
--
-- Tags 0-5 (6 constructors).
data ReasoningMode
  = Symbolic  -- ^ Symbolic (tag 0).
  | Neural  -- ^ Neural (tag 1).
  | SymToNeural  -- ^ SymToNeural (tag 2).
  | NeuralToSym  -- ^ NeuralToSym (tag 3).
  | Ensemble  -- ^ Ensemble (tag 4).
  | Cascade  -- ^ Cascade (tag 5).
  deriving (Show, Eq, Ord, Enum, Bounded)

-- | Convert a 'ReasoningMode' to its ABI tag value.
reasoningModeToTag :: ReasoningMode -> Word8
reasoningModeToTag = fromIntegral . fromEnum

-- | Decode a 'ReasoningMode' from its ABI tag value.
reasoningModeFromTag :: Word8 -> Maybe ReasoningMode
reasoningModeFromTag n
  | n <= fromIntegral (fromEnum (maxBound :: ReasoningMode)) = Just (toEnum (fromIntegral n))
  | otherwise = Nothing

-- ---------------------------------------------------------------------------
-- ProofStatus
-- ---------------------------------------------------------------------------

-- | Proof verification status.
--
-- Tags 0-5 (6 constructors).
data ProofStatus
  = Pending  -- ^ Pending (tag 0).
  | Attempting  -- ^ Attempting (tag 1).
  | Proved  -- ^ Proved (tag 2).
  | Failed  -- ^ Failed (tag 3).
  | Assumed  -- ^ Assumed (tag 4).
  | Vacuous  -- ^ Vacuous (tag 5).
  deriving (Show, Eq, Ord, Enum, Bounded)

-- | Convert a 'ProofStatus' to its ABI tag value.
proofStatusToTag :: ProofStatus -> Word8
proofStatusToTag = fromIntegral . fromEnum

-- | Decode a 'ProofStatus' from its ABI tag value.
proofStatusFromTag :: Word8 -> Maybe ProofStatus
proofStatusFromTag n
  | n <= fromIntegral (fromEnum (maxBound :: ProofStatus)) = Just (toEnum (fromIntegral n))
  | otherwise = Nothing

-- ---------------------------------------------------------------------------
-- ConstraintKind
-- ---------------------------------------------------------------------------

-- | Type constraint kinds.
--
-- Tags 0-7 (8 constructors).
data ConstraintKind
  = TypeEquality  -- ^ TypeEquality (tag 0).
  | Subtype  -- ^ Subtype (tag 1).
  | Linearity  -- ^ Linearity (tag 2).
  | Termination  -- ^ Termination (tag 3).
  | Totality  -- ^ Totality (tag 4).
  | Invariant  -- ^ Invariant (tag 5).
  | Refinement  -- ^ Refinement (tag 6).
  | DependentIndex  -- ^ DependentIndex (tag 7).
  deriving (Show, Eq, Ord, Enum, Bounded)

-- | Convert a 'ConstraintKind' to its ABI tag value.
constraintKindToTag :: ConstraintKind -> Word8
constraintKindToTag = fromIntegral . fromEnum

-- | Decode a 'ConstraintKind' from its ABI tag value.
constraintKindFromTag :: Word8 -> Maybe ConstraintKind
constraintKindFromTag n
  | n <= fromIntegral (fromEnum (maxBound :: ConstraintKind)) = Just (toEnum (fromIntegral n))
  | otherwise = Nothing

-- ---------------------------------------------------------------------------
-- NeuralBackend
-- ---------------------------------------------------------------------------

-- | Neural inference backend providers.
--
-- Tags 0-5 (6 constructors).
data NeuralBackend
  = LocalModel  -- ^ LocalModel (tag 0).
  | Claude  -- ^ Claude (tag 1).
  | Gemini  -- ^ Gemini (tag 2).
  | Mistral  -- ^ Mistral (tag 3).
  | Gpt  -- ^ GPT (tag 4).
  | CustomNeural  -- ^ CustomNeural (tag 5).
  deriving (Show, Eq, Ord, Enum, Bounded)

-- | Convert a 'NeuralBackend' to its ABI tag value.
neuralBackendToTag :: NeuralBackend -> Word8
neuralBackendToTag = fromIntegral . fromEnum

-- | Decode a 'NeuralBackend' from its ABI tag value.
neuralBackendFromTag :: Word8 -> Maybe NeuralBackend
neuralBackendFromTag n
  | n <= fromIntegral (fromEnum (maxBound :: NeuralBackend)) = Just (toEnum (fromIntegral n))
  | otherwise = Nothing

-- ---------------------------------------------------------------------------
-- Confidence
-- ---------------------------------------------------------------------------

-- | Inference confidence levels.
--
-- Tags 0-5 (6 constructors).
data Confidence
  = Verified  -- ^ Verified (tag 0).
  | HighNeural  -- ^ HighNeural (tag 1).
  | MediumNeural  -- ^ MediumNeural (tag 2).
  | LowNeural  -- ^ LowNeural (tag 3).
  | Unknown  -- ^ Unknown (tag 4).
  | Contradicted  -- ^ Contradicted (tag 5).
  deriving (Show, Eq, Ord, Enum, Bounded)

-- | Convert a 'Confidence' to its ABI tag value.
confidenceToTag :: Confidence -> Word8
confidenceToTag = fromIntegral . fromEnum

-- | Decode a 'Confidence' from its ABI tag value.
confidenceFromTag :: Word8 -> Maybe Confidence
confidenceFromTag n
  | n <= fromIntegral (fromEnum (maxBound :: Confidence)) = Just (toEnum (fromIntegral n))
  | otherwise = Nothing

-- ---------------------------------------------------------------------------
-- DriftKind
-- ---------------------------------------------------------------------------

-- | Knowledge drift types.
--
-- Tags 0-5 (6 constructors).
data DriftKind
  = NoDrift  -- ^ NoDrift (tag 0).
  | SemanticDrift  -- ^ SemanticDrift (tag 1).
  | ConfidenceDrift  -- ^ ConfidenceDrift (tag 2).
  | FactualDrift  -- ^ FactualDrift (tag 3).
  | TemporalDrift  -- ^ TemporalDrift (tag 4).
  | CatastrophicDrift  -- ^ CatastrophicDrift (tag 5).
  deriving (Show, Eq, Ord, Enum, Bounded)

-- | Convert a 'DriftKind' to its ABI tag value.
driftKindToTag :: DriftKind -> Word8
driftKindToTag = fromIntegral . fromEnum

-- | Decode a 'DriftKind' from its ABI tag value.
driftKindFromTag :: Word8 -> Maybe DriftKind
driftKindFromTag n
  | n <= fromIntegral (fromEnum (maxBound :: DriftKind)) = Just (toEnum (fromIntegral n))
  | otherwise = Nothing

-- ---------------------------------------------------------------------------
-- NeSyState
-- ---------------------------------------------------------------------------

-- | NeSy engine states.
--
-- Tags 0-5 (6 constructors).
data NeSyState
  = Idle  -- ^ Idle (tag 0).
  | Ready  -- ^ Ready (tag 1).
  | Reasoning  -- ^ Reasoning (tag 2).
  | Verifying  -- ^ Verifying (tag 3).
  | Drift  -- ^ Drift (tag 4).
  | Shutdown  -- ^ Shutdown (tag 5).
  deriving (Show, Eq, Ord, Enum, Bounded)

-- | Convert a 'NeSyState' to its ABI tag value.
neSyStateToTag :: NeSyState -> Word8
neSyStateToTag = fromIntegral . fromEnum

-- | Decode a 'NeSyState' from its ABI tag value.
neSyStateFromTag :: Word8 -> Maybe NeSyState
neSyStateFromTag n
  | n <= fromIntegral (fromEnum (maxBound :: NeSyState)) = Just (toEnum (fromIntegral n))
  | otherwise = Nothing

File: bindings/elixir/lib/proven_servers/nesy.ex
# SPDX-License-Identifier: MPL-2.0
# Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>

defmodule ProvenServers.Nesy do
  @moduledoc """
  NeSy types for the proven-servers ABI.
  
  Formally verified neurosymbolic AI types.
  Mirrors the Idris2 module `NesyABI.Types`.
  
  - `ReasoningMode` -- Neurosymbolic reasoning modes.
  - `ProofStatus` -- Proof verification status.
  - `ConstraintKind` -- Type constraint kinds.
  - `NeuralBackend` -- Neural inference backend providers.
  - `Confidence` -- Inference confidence levels.
  - `DriftKind` -- Knowledge drift types.
  - `NeSyState` -- NeSy engine states.
  
  All discriminant values match the Idris2 ABI tag definitions exactly.

  All tag values match the Idris2 ABI definitions exactly.
  """

  # ===========================================================================
  # ReasoningMode (tags 0-5)
  # ===========================================================================

  @typedoc """
  ReasoningMode types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type reasoning_mode ::
          :symbolic
          | :neural
          | :sym_to_neural
          | :neural_to_sym
          | :ensemble
          | :cascade

  @reasoning_mode_tags %{
    symbolic: 0,
    neural: 1,
    sym_to_neural: 2,
    neural_to_sym: 3,
    ensemble: 4,
    cascade: 5,
  }

  @tag_to_reasoning_mode Map.new(@reasoning_mode_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `ReasoningMode` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..5, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Nesy.reasoning_mode_from_tag(0)
      {:ok, :symbolic}
  """
  @spec reasoning_mode_from_tag(non_neg_integer()) :: {:ok, reasoning_mode()} | :error
  def reasoning_mode_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 5 do
    {:ok, Map.fetch!(@tag_to_reasoning_mode, tag)}
  end

  def reasoning_mode_from_tag(_tag), do: :error

  @doc """
  Encode a `ReasoningMode` to the C-ABI tag value.
  """
  @spec reasoning_mode_to_tag(reasoning_mode()) :: non_neg_integer()
  def reasoning_mode_to_tag(val) when is_map_key(@reasoning_mode_tags, val) do
    Map.fetch!(@reasoning_mode_tags, val)
  end

  @doc """
  All `ReasoningMode` variants in tag order.
  """
  @spec all_reasoning_modes() :: [reasoning_mode()]
  def all_reasoning_modes do
    [
      :symbolic, :neural, :sym_to_neural, :neural_to_sym, :ensemble,
      :cascade
    ]
  end

  # ===========================================================================
  # ProofStatus (tags 0-5)
  # ===========================================================================

  @typedoc """
  ProofStatus types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type proof_status :: :pending | :attempting | :proved | :failed | :assumed | :vacuous

  @proof_status_tags %{
    pending: 0,
    attempting: 1,
    proved: 2,
    failed: 3,
    assumed: 4,
    vacuous: 5,
  }

  @tag_to_proof_status Map.new(@proof_status_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `ProofStatus` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..5, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Nesy.proof_status_from_tag(0)
      {:ok, :pending}
  """
  @spec proof_status_from_tag(non_neg_integer()) :: {:ok, proof_status()} | :error
  def proof_status_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 5 do
    {:ok, Map.fetch!(@tag_to_proof_status, tag)}
  end

  def proof_status_from_tag(_tag), do: :error

  @doc """
  Encode a `ProofStatus` to the C-ABI tag value.
  """
  @spec proof_status_to_tag(proof_status()) :: non_neg_integer()
  def proof_status_to_tag(val) when is_map_key(@proof_status_tags, val) do
    Map.fetch!(@proof_status_tags, val)
  end

  @doc """
  All `ProofStatus` variants in tag order.
  """
  @spec all_proof_statuss() :: [proof_status()]
  def all_proof_statuss, do: [:pending, :attempting, :proved, :failed, :assumed, :vacuous]

  # ===========================================================================
  # ConstraintKind (tags 0-7)
  # ===========================================================================

  @typedoc """
  ConstraintKind types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type constraint_kind ::
          :type_equality
          | :subtype
          | :linearity
          | :termination
          | :totality
          | :invariant
          | :refinement
          | :dependent_index

  @constraint_kind_tags %{
    type_equality: 0,
    subtype: 1,
    linearity: 2,
    termination: 3,
    totality: 4,
    invariant: 5,
    refinement: 6,
    dependent_index: 7,
  }

  @tag_to_constraint_kind Map.new(@constraint_kind_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `ConstraintKind` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..7, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Nesy.constraint_kind_from_tag(0)
      {:ok, :type_equality}
  """
  @spec constraint_kind_from_tag(non_neg_integer()) :: {:ok, constraint_kind()} | :error
  def constraint_kind_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 7 do
    {:ok, Map.fetch!(@tag_to_constraint_kind, tag)}
  end

  def constraint_kind_from_tag(_tag), do: :error

  @doc """
  Encode a `ConstraintKind` to the C-ABI tag value.
  """
  @spec constraint_kind_to_tag(constraint_kind()) :: non_neg_integer()
  def constraint_kind_to_tag(val) when is_map_key(@constraint_kind_tags, val) do
    Map.fetch!(@constraint_kind_tags, val)
  end

  @doc """
  All `ConstraintKind` variants in tag order.
  """
  @spec all_constraint_kinds() :: [constraint_kind()]
  def all_constraint_kinds do
    [
      :type_equality, :subtype, :linearity, :termination, :totality,
      :invariant, :refinement, :dependent_index
    ]
  end

  # ===========================================================================
  # NeuralBackend (tags 0-5)
  # ===========================================================================

  @typedoc """
  NeuralBackend types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type neural_backend :: :local_model | :claude | :gemini | :mistral | :gpt | :custom_neural

  @neural_backend_tags %{
    local_model: 0,
    claude: 1,
    gemini: 2,
    mistral: 3,
    gpt: 4,
    custom_neural: 5,
  }

  @tag_to_neural_backend Map.new(@neural_backend_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `NeuralBackend` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..5, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Nesy.neural_backend_from_tag(0)
      {:ok, :local_model}
  """
  @spec neural_backend_from_tag(non_neg_integer()) :: {:ok, neural_backend()} | :error
  def neural_backend_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 5 do
    {:ok, Map.fetch!(@tag_to_neural_backend, tag)}
  end

  def neural_backend_from_tag(_tag), do: :error

  @doc """
  Encode a `NeuralBackend` to the C-ABI tag value.
  """
  @spec neural_backend_to_tag(neural_backend()) :: non_neg_integer()
  def neural_backend_to_tag(val) when is_map_key(@neural_backend_tags, val) do
    Map.fetch!(@neural_backend_tags, val)
  end

  @doc """
  All `NeuralBackend` variants in tag order.
  """
  @spec all_neural_backends() :: [neural_backend()]
  def all_neural_backends, do: [:local_model, :claude, :gemini, :mistral, :gpt, :custom_neural]

  # ===========================================================================
  # Confidence (tags 0-5)
  # ===========================================================================

  @typedoc """
  Confidence types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type confidence ::
          :verified
          | :high_neural
          | :medium_neural
          | :low_neural
          | :unknown
          | :contradicted

  @confidence_tags %{
    verified: 0,
    high_neural: 1,
    medium_neural: 2,
    low_neural: 3,
    unknown: 4,
    contradicted: 5,
  }

  @tag_to_confidence Map.new(@confidence_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `Confidence` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..5, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Nesy.confidence_from_tag(0)
      {:ok, :verified}
  """
  @spec confidence_from_tag(non_neg_integer()) :: {:ok, confidence()} | :error
  def confidence_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 5 do
    {:ok, Map.fetch!(@tag_to_confidence, tag)}
  end

  def confidence_from_tag(_tag), do: :error

  @doc """
  Encode a `Confidence` to the C-ABI tag value.
  """
  @spec confidence_to_tag(confidence()) :: non_neg_integer()
  def confidence_to_tag(val) when is_map_key(@confidence_tags, val) do
    Map.fetch!(@confidence_tags, val)
  end

  @doc """
  All `Confidence` variants in tag order.
  """
  @spec all_confidences() :: [confidence()]
  def all_confidences do
    [
      :verified, :high_neural, :medium_neural, :low_neural, :unknown,
      :contradicted
    ]
  end

  # ===========================================================================
  # DriftKind (tags 0-5)
  # ===========================================================================

  @typedoc """
  DriftKind types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type drift_kind ::
          :no_drift
          | :semantic_drift
          | :confidence_drift
          | :factual_drift
          | :temporal_drift
          | :catastrophic_drift

  @drift_kind_tags %{
    no_drift: 0,
    semantic_drift: 1,
    confidence_drift: 2,
    factual_drift: 3,
    temporal_drift: 4,
    catastrophic_drift: 5,
  }

  @tag_to_drift_kind Map.new(@drift_kind_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `DriftKind` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..5, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Nesy.drift_kind_from_tag(0)
      {:ok, :no_drift}
  """
  @spec drift_kind_from_tag(non_neg_integer()) :: {:ok, drift_kind()} | :error
  def drift_kind_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 5 do
    {:ok, Map.fetch!(@tag_to_drift_kind, tag)}
  end

  def drift_kind_from_tag(_tag), do: :error

  @doc """
  Encode a `DriftKind` to the C-ABI tag value.
  """
  @spec drift_kind_to_tag(drift_kind()) :: non_neg_integer()
  def drift_kind_to_tag(val) when is_map_key(@drift_kind_tags, val) do
    Map.fetch!(@drift_kind_tags, val)
  end

  @doc """
  All `DriftKind` variants in tag order.
  """
  @spec all_drift_kinds() :: [drift_kind()]
  def all_drift_kinds do
    [
      :no_drift, :semantic_drift, :confidence_drift, :factual_drift,
      :temporal_drift, :catastrophic_drift
    ]
  end

  # ===========================================================================
  # NeSyState (tags 0-5)
  # ===========================================================================

  @typedoc """
  NeSyState types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type ne_sy_state :: :idle | :ready | :reasoning | :verifying | :drift | :shutdown

  @ne_sy_state_tags %{
    idle: 0,
    ready: 1,
    reasoning: 2,
    verifying: 3,
    drift: 4,
    shutdown: 5,
  }

  @tag_to_ne_sy_state Map.new(@ne_sy_state_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `NeSyState` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..5, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Nesy.ne_sy_state_from_tag(0)
      {:ok, :idle}
  """
  @spec ne_sy_state_from_tag(non_neg_integer()) :: {:ok, ne_sy_state()} | :error
  def ne_sy_state_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 5 do
    {:ok, Map.fetch!(@tag_to_ne_sy_state, tag)}
  end

  def ne_sy_state_from_tag(_tag), do: :error

  @doc """
  Encode a `NeSyState` to the C-ABI tag value.
  """
  @spec ne_sy_state_to_tag(ne_sy_state()) :: non_neg_integer()
  def ne_sy_state_to_tag(val) when is_map_key(@ne_sy_state_tags, val) do
    Map.fetch!(@ne_sy_state_tags, val)
  end

  @doc """
  All `NeSyState` variants in tag order.
  """
  @spec all_ne_sy_states() :: [ne_sy_state()]
  def all_ne_sy_states, do: [:idle, :ready, :reasoning, :verifying, :drift, :shutdown]

end
